community/detectors/localai_cve_2024_2029/src/main/java/com/google/tsunami/plugins/detectors/rce/cve20242029/LocalAiCve20242029RceDetector.java

```diff
@@ -0,0 +1,251 @@
+import com.google.tsunami.proto.Vulnerability;
+import com.google.tsunami.proto.VulnerabilityId;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import javax.inject.Inject;
+
+/** A {@link VulnDetector} that detects exposed LocalAI API server. */
+@PluginInfo(
+    type = PluginType.VULN_DETECTION,
+
+    // name of the plugin
+    name = "ExposedLocalAIDetector",
+    version = "0.1",
+
+    // detailed description of the plugin
+    description =
+        "This plugin detects exposed LocalAI API serve that vulnerable to cve-2024-2029."
+            + "this CVE allows attacker to inject arbitrary OS commands "
+            + "during a file upload upload by a POST HTTP request",
+    author = "TheVampKid",
+    bootstrapModule = LocalAiCve20242029RceDetectorBootstrapModule.class)
+public final class LocalAiCve20242029RceDetector implements VulnDetector {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  private final PayloadGenerator payloadGenerator;
+  private final Clock utcClock;
+  private final HttpClient httpClient;
+  private final int oobSleepDuration;
+
+  @Inject
+  LocalAiCve20242029RceDetector(
+      HttpClient httpClient,
+      @UtcClock Clock utcClock,
+      PayloadGenerator payloadGenerator,
+      @OobSleepDuration int oobSleepDuration) {
+    this.httpClient = checkNotNull(httpClient).modify().setFollowRedirects(true).build();
+    this.utcClock = checkNotNull(utcClock);
+    this.payloadGenerator = checkNotNull(payloadGenerator);
+    this.oobSleepDuration = oobSleepDuration;
+  }
+
+  @Override
+  public DetectionReportList detect(
+      TargetInfo targetInfo, ImmutableList<NetworkService> matchedServices) {
+    logger.atInfo().log("Starting LocalAI CVE-2024-2029 RCE detection.");
+
+    return DetectionReportList.newBuilder()
+        .addAllDetectionReports(
+            matchedServices.stream()
+                .filter(NetworkServiceUtils::isWebService)
+                .filter(this::checkIfLocalAiWebService)
+                .filter(this::isServiceVulnerable)
+                .map(networkService -> buildDetectionReport(targetInfo, networkService))
+                .collect(toImmutableList()))
+        .build();
+  }
+
+  private boolean checkIfLocalAiWebService(NetworkService networkService) {
+    String targetUrl = NetworkServiceUtils.buildWebApplicationRootUrl(networkService);
+
+    String modelsUrl = targetUrl + "models";
+    HttpResponse response;
+    try {
+      response = httpClient.send(get(modelsUrl).withEmptyHeaders().build(), networkService);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    if (response.bodyJson().isEmpty()) {
+      return false;
+    }
+    return !response.bodyJson().get().getAsJsonObject().get("object").getAsString().isEmpty();
+  }
+
+  private boolean isServiceVulnerable(NetworkService networkService) {
```

Comment:
Please rewrite this function, you can use `HttpClient` to build the request without manually crafting it, this is both for better readability and code robustness.